worker/provisioner/containerprovisioner_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provisioner_test
+
+import (
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/juju/errors"
+	jc "github.com/juju/testing/checkers"
+	"github.com/juju/utils/arch"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1/workertest"
+
+	"github.com/juju/juju/container/kvm/mock"
+	kvmtesting "github.com/juju/juju/container/kvm/testing"
+	"github.com/juju/juju/core/instance"
+	supportedversion "github.com/juju/juju/juju/version"
+	"github.com/juju/juju/state"
+	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/worker/provisioner"
+)
+
+type kvmProvisionerSuite struct {
+	CommonProvisionerSuite
+	kvmtesting.TestSuite
+
+	events chan mock.Event
+}
+
+var _ = gc.Suite(&kvmProvisionerSuite{})
+
+func (s *kvmProvisionerSuite) SetUpSuite(c *gc.C) {
+	if runtime.GOOS == "windows" {
+		c.Skip("Skipping kvm tests on windows")
+	}
+	s.CommonProvisionerSuite.SetUpSuite(c)
+	s.TestSuite.SetUpSuite(c)
+}
+
+func (s *kvmProvisionerSuite) TearDownSuite(c *gc.C) {
+	s.TestSuite.TearDownSuite(c)
+	s.CommonProvisionerSuite.TearDownSuite(c)
+}
+
+func (s *kvmProvisionerSuite) SetUpTest(c *gc.C) {
+	s.CommonProvisionerSuite.SetUpTest(c)
+	s.TestSuite.SetUpTest(c)
+
+	s.events = make(chan mock.Event, 25)
+	s.ContainerFactory.AddListener(s.events)
+}
+
+func (s *kvmProvisionerSuite) nextEvent(c *gc.C) mock.Event {
+	select {
+	case event := <-s.events:
+		return event
+	case <-time.After(coretesting.LongWait):
+		c.Fatalf("no event arrived")
+	}
+	panic("not reachable")
+}
+
+func (s *kvmProvisionerSuite) expectStarted(c *gc.C, machine *state.Machine) string {
+	s.State.StartSync()
+	event := s.nextEvent(c)
+	c.Assert(event.Action, gc.Equals, mock.Started)
+	err := machine.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	s.waitInstanceId(c, machine, instance.Id(event.InstanceId))
+	return event.InstanceId
+}
+
+func (s *kvmProvisionerSuite) expectStopped(c *gc.C, instId string) {
+	s.State.StartSync()
+	event := s.nextEvent(c)
+	c.Assert(event.Action, gc.Equals, mock.Stopped)
+	c.Assert(event.InstanceId, gc.Equals, instId)
+}
+
+func (s *kvmProvisionerSuite) expectNoEvents(c *gc.C) {
+	select {
+	case event := <-s.events:
+		c.Fatalf("unexpected event %#v", event)
+	case <-time.After(coretesting.ShortWait):
+		return
+	}
+}
+
+func (s *kvmProvisionerSuite) TearDownTest(c *gc.C) {
+	close(s.events)
+	s.TestSuite.TearDownTest(c)
+	s.CommonProvisionerSuite.TearDownTest(c)
+}
+
+func (s *kvmProvisionerSuite) newKvmProvisioner(c *gc.C) provisioner.Provisioner {
```

Comment:
Cover container provisioner unit tests.  Use a mock broker to as the KVM broker to start it.